### ISSUE TITLE
[Suggestion] CRT effect (a new and badass one)

### ISSUE THREAD

**Issue by SilentMRG** [finalburnneo/FBNeo#1897]
Hello everyone!

This CRT filter is called "Scan lines of CRT". The screenshot below is from the latest version of PPSSPP, which has two CRT filters, one static, and one animated, which is the one in the screenshot.

![ms3](https://github.com/user-attachments/assets/59047d3e-4193-45fa-a65a-9ec52e6643fc)

Anyway, I play in PPSSPP using DirectX 9, and I saw that FBN has support for this backend, so I was thinking that adding this filter would be cool, as it is a very beautiful and useful CRT filter for 2D games. Yes, I know FBN has its own CRT filter, but in my humble opinion it is too basic, with a generic appearance.

Thank you for your attention...
Best regards!

**Comment by barbudreadmon** [comment]
Filters in FBNeo aren't hardware-rendered. I believe computing complex filters like this using only the cpu would have a huge performance hit.

**Comment by dinkc64** [comment]
We support dx9 hardware shaders (D3DXCreateEffectFromFile() - whatever format this is?  some sort of hlsl?)
So, where can one find this filter?  I took a very quick look at ppsspp code and couldn't find it.  

best regards,
- dink

**Comment by SilentMRG** [comment]
> We support dx9 hardware shaders (D3DXCreateEffectFromFile() - whatever format this is? some sort of hlsl?) So, where can one find this filter? I took a very quick look at ppsspp code and couldn't find it.
> 
> best regards,
> 
>     * dink

Hey dink, I think this is the one:

![Sem título](https://github.com/user-attachments/assets/0fbecc66-af11-4242-810e-94cdf78aec88)

Note that it is in "assets/shaders", there are many filter files, however, the only one called "scanlines" is this one. Maybe it works in conjunction with another filter? The problem will be guessing the name of the other file.

**Comment by SilentMRG** [comment]
To make it a little easier, here it is: [scanlines.zip](https://github.com/user-attachments/files/17468871/scanlines.zip)

**Comment by SilentMRG** [comment]
Edit... Here's something that might help: https://github.com/hrydgard/ppsspp/blob/master/assets/shaders/scanlines.fsh

**Comment by barbudreadmon** [comment]
@dinkc64 Oh, i totally forgot we had an external shader loading sort of thing.
Maybe something like https://github.com/septag/glslcc could help converting their glsl shaders into an hlsl one ?

**Comment by dinkc64** [comment]
Thanks guys!
I can't make that tool help at all, but, feel free to try :)

best regards,
- dink

**Comment by SilentMRG** [comment]
Hey guys, nice people said things that might be helpful: https://github.com/hrydgard/ppsspp/issues/19548

**Comment by dinkc64** [comment]
Thanks guys for the help!
I ended up porting it by hand with Microsoft's glsl -> hlsl docu, so far it looks like this:
(note, I have enabled linear interpolation for a little blurring effect)
![test](https://github.com/user-attachments/assets/a5094370-bcb5-4f64-9c4d-094db6aa55cc)

**Comment by dinkc64** [comment]
can you tell me which metal slug is in the first pic, so I can match the scene exactly to get everything just right?

**Comment by barbudreadmon** [comment]
I think that's mslug3 ? I sort of remember you start from a beach in that one.

**Comment by SilentMRG** [comment]
It's the first mission in Metal Slug 3. =)
Marco is at the beginning of the mission, after parachuting.

dink, the effect is very similar, that's good! But just one question: Did you add the "distortion" effect? I mean, there's a big horizontal scan line that goes up the screen in a vertical direction. If you enlarge the image I posted, you will be able to see it in the dark corners of the image. A second alternative is to observe this animated scan line effect in PPSSPP itself. In PPSSPP there is also a kind of "timer" to adjust and make this animated line go faster or slower, from what I could see.

Thank you mates!

**Comment by dinkc64** [comment]
SilentMRG can you tell me the values for this shader's setting you use?

**Comment by dinkc64** [comment]
Also maybe make a video? I really don't want to have to install this emulator and find roms, etc. :)
Besides, a lot of apps these days don't like my outdated OS.

**Comment by barbudreadmon** [comment]
note : i tried to find a video of this on youtube the other day, but i couldn't, and everything i found was too low quality to see anything anyway (it might be due to youtube compression ?)

**Comment by dinkc64** [comment]
barbudreadmon, i sent you a message somewhere, try to find it

**Comment by SilentMRG** [comment]
Mates, before anything else, notice that I present two CRT filters in the video: the "animated" one and then briefly the "non-animated" one. I use the default values ​​for the animated filter, but in the video I change the values ​​so you can see how everything works in real time.

Anyway, here's the sampling video: https://www.youtube.com/watch?v=MrPY_amlyrc
Unfortunately, it was not possible to record the audio, but that is irrelevant considering the video purpose.

I used this program to record: https://github.com/ShareX/ShareX
And to cut some frames I used this: https://github.com/mltframework/shotcut
Great projects, by the way.

I hope the video helps you understand the whole thing.

Best regards to both of you!

**Comment by dinkc64** [comment]
SilentMRG, thanks!  That helps :)

btw, I was trying to get barbudreadmon's attention re: another issue above there :)

**Comment by SilentMRG** [comment]
dinkc64, about the animated filter, I noticed that it first uses a bilinear filter + motion blur + scan lines (which also seem to be animated) and lastly it applies the animation that big scan line. =)

**Comment by dinkc64** [comment]
Some things...

I'm not sure if I got the timing right for the animation thing, please let me know if it needs to go faster or slower? :)

Hmm, I don't know the source of the bilinear filter, because I don't see it in the filter/shader file.  You could just use the linear filter option of the dx9 blitter, though!

For the meantime you'll have to edit the setting manually in the shader file until we make some sort of UI control for it.   I can't guarantee that'll ever happen, though.  (I really hate UI stuff)
But, just right click on the game screen twice to reload the filter. after editing it, of course :)
Change the numbers on this line:
const float setting = 0.00;

[crt_retrosl.zip](https://github.com/user-attachments/files/17501740/crt_retrosl.zip)

Just to be clear, this ported filter isn't being distributed by our project (just like the others) due to licensing conflicts and whatnot.

**Comment by dinkc64** [comment]
Some things...

I'm not sure if I got the timing right for the animation thing, please let me know if it needs to go faster or slower? :)

Hmm, I don't know the source of the bilinear filter, because I don't see it in the filter/shader file.  You could just use the linear filter option of the dx9 blitter, though!

For the meantime you'll have to edit the setting manually in the shader file until we make some sort of UI control for it.   I can't guarantee that'll ever happen, though.  (I really hate UI stuff)
But, just right click on the game screen twice to reload the filter. after editing it, of course :)
Change the numbers on this line:
const float setting = 0.00;

[crt_retrosl.zip](https://github.com/user-attachments/files/17501740/crt_retrosl.zip)

**Comment by dinkc64** [comment]
Working on a solution to change the shader parameters via the ui, fingers crossed all goes well :)

**Comment by dinkc64** [comment]
Working on a solution to change the shader parameters via the ui, fingers crossed all goes well :)

**Comment by SilentMRG** [comment]
dinkc64, thanks so far for all the work!

But, I must be missing something... I did it the way you said, and I'm getting this: 
![Sem título](https://github.com/user-attachments/assets/4b40e1a0-04a2-4651-a3a9-bf2cc421ad4a)

**Comment by SilentMRG** [comment]
Forget my drunk post above, I'm an idiot. The error is due to the lack of other shaders, hehe. Downloading the other shaders solved the problem.

What I can't find is the name of the filter in the filter list, see...

![Sem título](https://github.com/user-attachments/assets/8a1d99d7-37be-482d-bdfe-009c225be46b)

Would the tentative option be "none"? When I select it I get a shader, but not animated, even changing the values ​​in the file.

**Comment by dinkc64** [comment]
Hi SilentMRG,
Make sure to get the updated fbneo exe from here:
https://github.com/finalburnneo/FBNeo/releases/tag/latest

It'll add the new shader to the list :)

best regards,
- dink

**Comment by SilentMRG** [comment]
Also forget that other drunk post I made above. I was using an old nightly build, upgrading to the latest one resolved the issue.

Well, after several tests I did, the following value for "const float setting = 0.030" is the best, however, it only looks good on full screen, because in a window the scan lines seem bigger and kind of misaligned, on full screen , the speed and the general appearance is fine. It seems that the value mentioned above behaves differently in windowed mode, just wondering. 

One question: Is there a way to increase the sharpness/brightness?

**Comment by SilentMRG** [comment]
I forgot to say that the value 0.030 is an approximate value of the value 0.349 in PPSSPP. The default value in the latter is 1.002, this in FBN becomes very fast.

**Comment by SilentMRG** [comment]
I must admit that I was doing one thing wrong: Using the 4x4 (1080p) CRT filter together with the Retro Scanlines filter, so the lines were messed up in windowed mode. 

A curious fact is that using other filters in combination with the Retro filter interferes with the speed of the lines, either more or less, so the value in "const float setting" will need to be modified depending on the combination used.

Getting straight to the point, the correct thing is: Double pixels (3D hardware).
Well, with that in mind, the best way to use the CRT Retro filter is with this combination:

- Linear filtering;
- Double pixels (3D hardware);
- With "const float setting = 0.020" or a higher or lower value, anyway, this will depend on the player's taste.

With these combinations, the Retro scanning lines will be side by side with the PPSSPP filter, the difference between them is minimal, so minimal that only a boring person like me would be able to notice it.

You did it, dinkc64, you did it! I'll buy you a coffee when I can. =D

Thank you very much for everything... 
Best regards!

**Comment by dinkc64** [comment]
Great, glad it worked out! :)
 I see you've noticed that mixing filters isn't always a good thing.. :)

I'm particularly interested in the minimal differences, because it would be nice to get things 100% right, if you're interested to talk about it.  Would you care to link all of the stuff needed to run ppsspp & metal slug with that filter?  I would like to do the comparison on my side, but, I don't want to have to chase files all night long to get it working XD

best regards,
- dink

**Comment by SilentMRG** [comment]
So, the first difference I noticed is in the larger scan line, the one that crosses the entire screen in a vertical direction, but the line is horizontal, of course. In PPSSPP this line takes up the entire screen, I mean you have the line going up to the side black bars, and in FBN the line only goes up inside the game screen, that is, in the 4:3 ratio.

The second difference is precisely in the speed, in PPSSPP the standard speed is not so fast and in FBN, if I use this standard speed it will be extremely fast, and it was difficult for me to reach a speed similar to that of PPSSPP. 

And third and last, maybe it's a problem with my vision, but I could see that the filter in PPSSPP slightly changes the color palette to a colder tone, and in FBN I didn't notice this difference. If you look at that video I posted, you will see that the other CRT filter changes the palette to a greenish tone, what I mean is that in PPSSPP it seems that the filters slightly modify the colors of the games, but like I said, maybe it's the my vision only, hehe.

As for running Metal Slug Anthology on your side, all you need is the latest release of PPSSPP (portable version): https://www.ppsspp.org/
Use default video settings in full screen mode.

The file, you know, can be obtained from a site called free roms. If you don't mind, I'll post the direct link here. =)
I don't know if that's what you meant when you said "Would you care to link all of the stuff needed to run ppsspp & metal slug with that filter?" If that wasn't the case, apologies in advance.

**Comment by SilentMRG** [comment]
It's very good with another game: https://www.youtube.com/watch?v=Re4DUFOVTbs
Unfortunately, I was unable to record in full screen. So expect square video with a maximum quality of old-school 480p. XD

**Comment by SilentMRG** [comment]
Now that I saw it! There is another CRT file inside the PPSSPP shaders folder. Take a look: 
![Sem título](https://github.com/user-attachments/assets/8989227f-b937-49b0-af0c-aff89195198e)

Note that it is larger than the other file in the other image I posted above. I think this is the animated CRT filter shader, the other smaller one should be the non-animated scanline filter.

**Comment by dinkc64** [comment]
Thanks for the info's!  Will look into things tomorrow :)
btw: crt.fsh is the one I ported. the scanlines.fsh one ended up not being the one you wanted, so I found the right one.

also, did you check out "crt cga"?  This is my favorite shader

best regards,
- dink

**Comment by SilentMRG** [comment]
Hello dinkc64!

I hope the information helps in some way. I just want to add a detail that I forgot: In the PPSSPP video settings it is marked to run in OpenGL, choose Direct3D 9, it was using this backend that I recorded the video.

About the "crt cga" filter, I'll test it later... Thanks for the tip! =)

Best regards!

**Comment by SilentMRG** [comment]
Hey dinkc64,

I have a question: Is it normal for scan lines to be vertical? I think they should be horizontal.

![Sem título](https://github.com/user-attachments/assets/de42e33b-8c11-4486-923d-9100351b90e3)

Regarding the filter "crt cga" you indicated, I didn't find it in the list...

![Sem título0](https://github.com/user-attachments/assets/59249a03-2abc-41b5-add5-2fcb48ad787f)

Would it be the "CGWG Fast" filter?

Thanks in advance!
And I'm sorry for bothering you with this.

**Comment by dinkc64** [comment]
Hi @SilentMRG,
Thanks for the tips for PPSSPP!

It's normal about the vertical scanlines, because on vertical games, the monitor is actually rotated 90degrees :)

About the missing cga filter: I'm afraid the UI translation will need to be updated for all the languages, for now revert it to the default language (English), and it will be OK.  

It's not a bother, so don't worry! :)

best regards,
- dink

**Comment by dinkc64** [comment]
So, the good news:
I got the filter to match more like in ppsspp, but some games the animation will be slightly slower or faster.  This is due to the number of vertical lines that the emulated game has - in this shader the animation speed is entirely dependent on the vertical scanlines of the game.  
The UI menu for changing this value is almost done, also :)
I think it should be ready to test soon, I'll post back here when it's ready

best regards,
- dink

**Comment by SilentMRG** [comment]
Thank you for your kindness and work!

**Comment by dinkc64** [comment]
Welcome!

**Comment by dinkc64** [comment]
Hi,
If everything goes well, the latest nightly exe should have the settings menu for hardfx (shaders).  So far the only shader with configurable stuff is the "crt retro scanlines".  Also you'll need the updated shader file:
[retrosl.zip](https://github.com/user-attachments/files/17552120/retrosl.zip)

best regards,
- dink

**Comment by SilentMRG** [comment]
Hello... =)

I just tested these games here:

Metal Slug;
Shock Troopers;
Alien vs. Predator.

In terms of speed, I set it to maximum and I didn't notice any slowdown or very quickly in these games, they ran at the same speed, as if they were without the filter. I also noticed that in the other version of the filter there was a larger scanning line that went up the screen horizontally, but in this new version it is absent, anyway, this is just a detail that doesn't make much difference; and yes, the general appearance and the way the smaller scan lines behave is very similar to the PPSSPP version. =D

I was already forgetting, I tested the cga filter and on my side it was slow as hell! XD a pity ... I liked his appearance, especially the details in the corners, imitating that oval screen aspect of the CRT.

Thanks again for all the good work...
Best regards!

**Comment by SilentMRG** [comment]
Another thing I noticed now, if I put the value 0.50 I will have fewer scan lines, and a slower animation (obviously), but if I put the maximum value of 2.00 I will have more scan lines and a faster animation.

Just a suggestion: It would be nice if the slider just changed the animation speed without messing with the amount of scan lines, the value of 2.00 presents a good amount of scan lines, but, I wanted that amount of scan lines when I would use, for example, 0.50 speed. I don't know if I was clear with the information, hehe. I remember that in the other version of the filter, changing the speed only changed the speed, without touching the number of scan lines, so it would be interesting to add an extra slider just for the number of scan lines.

**Comment by dinkc64** [comment]
Hi SilentMRG,
I tried the filter with Metal Slug 3 - If do .50 or 2.0, it has the same amount of scanlines, But keep in mind that the animation is tied to this value, so it might move up or down a little bit depending on the value here.  So when you're moving the slider, you'll see the "position in the animation sequence" changes.

Regarding the lovely cga filter, it needs a beastly video card to use it fullscreen.  I just use it windowed because  have onboard video.

best regards,
- dink

**Comment by dinkc64** [comment]
Just wanted to let you know, the "Portuguese (Brazil).flt" translation has been updated, thanks to Igor_Arabe!

**Comment by SilentMRG** [comment]
Hello dinkc64, 
I understand... For me the number of scan lines decreased according to the slider value, I didn't know that in PPSSPP the filter was always animated. So, no problem. =)

That's what I realized! I also have an iGPU and just windowed mode solved the issue, but it didn't resolve it 100%, I also can't expect much from an Intel HD Graphics. XD

Then I will look at the location in PT-BR. Thanks for letting me know! :)

Dude, I think that's all... Thank you from the bottom of my heart for your kindness, work and attention so far! I always liked FBA and when I discovered FBN I started using it, and I didn't know there were cool people behind the project, this will be another reason that will lead me to always use FBN. ^^

Now let's turn our attention to this: https://github.com/finalburnneo/FBNeo/issues/1359#issuecomment-1802577022
It's a joke! XD

Well, I'll close this.

All the best always...
Best regards!
